theano/tensor/basic.py

```diff
@@ -5941,44 +5991,136 @@ def infer_shape(self, node, shapes):
+    vector : symbolic vector
+        A vector representing the diagonal elements.
+
+    """
     return Diagonal(offset, axis1, axis2)(a)
 
 
 class Diag(Op):
+    """
+    An op that copies a vector to the diagonal of an empty matrix. It does the
+    inverse of Diagonal.
 
-    __props__ = ()
+    Usage: T.diag()(x)
+
+    `x` should be a tensor vector. The parenthesis in the front should indicate
+    which main diagonal the vector value goes into. By default it is set to (
+    `0`, which corresponds to setting the values of x to the main diagonal in
+    the returned matrix. Currently the gradient is valid only when `offset=0`.
+
+    Parameters
+    ----------
+    offset : int
+        Indicates which diagonal to put `x` into. Defaults to `0`.
+
+    x: symbolic vector
+        A tensor vector consists of diagonal values.
+
+    Returns
+    -------
+    tensor : symbolic tenstor
+        A tensor with passed vector values at its corresponding diagonal.
+
+    """
+    __props__ = ("offset", )
+    default_offset = 0
+
+    def __init__(self, offset=0):
+        if numpy_diagonal_return_view:
+            self.view_map = {0: [0]}
+        self.offset = offset
 
     def make_node(self, diag):
         diag = as_tensor_variable(diag)
         if diag.type.ndim != 1:
             raise TypeError('data argument must be a vector', diag.type)
-
         return Apply(self, [diag], [matrix(dtype=diag.dtype)])
 
     def perform(self, node, inputs, outputs):
         (z,) = outputs
-        z[0] = numpy.diag(inputs[0])
+        z[0] = numpy.diag(inputs[0], self.offset)
 
     def grad(self, inputs, gout):
         (gz,) = gout
-        return [diagonal(gz)]
+        if self.has_default_props():
+            return [diagonal(gz)]
+        else:
+            return [grad_not_implemented(self, 0, inputs[0])]
 
     def infer_shape(self, nodes, shapes):
         return [(shapes[0][0],) * 2]
 
+    def has_default_props(self):
+        if self.offset == self.default_offset:
+            return True
+        else:
+            return False
+
 
 def diag(v, k=0):
+    """
+    A helper function for two ops: theano.tensor.Diagonal and
+    theano.tensor.Diag. It both accepts tensor vector and tensor matrix. While
+    the passed tensor variable `v` has `v.ndim>=2`, it builds a Diagonal
+    instance, and returns a vector with its entries equal to `v`'s main
+    diagonal; otherwise if `v.ndim` is `1`, it builds a Diag instance, and
+    returns a matrix with `v` at its k-th diaogonal.
+
+    Parameters
+    ----------
+    v : symbolic tensor
+    k : int
+        offset
+
+    Returns
+    -------
+    tensor : symbolic tensor
+
+    """
     if v.ndim == 1:
         assert k == 0, "diagonals other than main are not implemented"
         return Diag()(v)
-    elif v.ndim == 2:
+    elif v.ndim >= 2:
         return diagonal(v, k)
     else:
-        raise ValueError("Input must be 1- or 2-d.")
+        raise TypeError("Input must has v.ndim >=1.")
```

Comment:
In the test code we can find that all confusions between vector and matrix are treated as TypeError, it looks not natural to have a ValueError in the middle of many TypeErrors.